### PR TITLE
Wrap qiskit.Aer and qiskit.IBMQ with lazy loading object

### LINKED ISSUE #5532: Avoid potential import cycles from Aer and IBMQ
### What is the expected enhancement?

Right now the effort to remove namespace packaging in #5089 is blocked on the modules level imports of `Aer` and `IBMQ`. This is because this causes an import cycle when those factory objects are no longer in the qiskit namespace. Basically, `qiskit_aer` imports `qiskit` whichi imports `qiskit_aer.Aer` which can't be resolved by python. To move forward on this we need to make these objects lazily load `qiskit.providers.aer`/`qiskit_aer` and `qiskit.providers.ibmq`/`qiskit_ibmq_provider`. The easiest way to handle this is to wait for python 3.7 to be the minimum python version and then we can just use a module level `__getattr__` and import when `Aer` or `IBMQ` are accessed. But that is likely 6+ months away since 0.17.0 will start the deprecation timer on Python 3.6 support. In the meantime we should be able do something like:

```python3
class _AerWrapper:

    def __init__(self):
       self.aer = None
       
    def __getattr__(self, attr):
       if self._aer is None:
           try:
                from qiskit.providers.aer import Aer
            except ImportError:
                raise ImportError("useful message")
            self.aer = Aer
      return getattr(self.aer, attr)
```
(note this is completely untested, just my rough idea).

This should hopefully enable us to remove the import cycle at the module level. It would also have a noticeable benefit for #5100 as we're not unconditionally importing ibmq and aer all the time.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit migrates the qiskit.Aer and qiskit.IBMQ module attributes to
be lazy loading instances of lazy loading wrapper classes. The intent
here is to avoid importing from qiskit-aer or qiskit-ibmq-provider from
qiskit-terra, while this is safe while the packages share a shared
namespace we've been actively working to remove the use of namespace
packaging (see #5089, #4767, and Qiskit/qiskit#559) and the circular
dependency caused by re-exporting these attributes is blocking progress
on this. By using a lazy loading wrapper class we avoid an import type
circular dependency and opportunistically use qiskit-aer and/or
qiskit-ibmq-provider at runtime only if they're present after everything
is imported. This also may have some benefit for the overall import
performance of qiskit (being tracked in #5100) as it removes qiskit-aer
and qiskit-ibmq-provider from the import path unless they're being used.

### Details and comments

Fixes #5532

### CODE REVIEW

**Comment by nonhermitian** [comment]
I thought these preloading of modules was to be removed. Are these to be removed later?

**Comment by mtreinish** [comment]
> I thought these preloading of modules was to be removed. Are these to be removed later?

My thinking was that this is a prerequisite for a deprecation and eventual removal of the attributes (it will probably be a long deprecation because these are widely used today). This gives us a place to raise a deprecation warnings (which wouldn't be an option otherwise while we still support py3.6) in the future while also fixing a potential import cycle today so we can move forward on removing the namespace packages..

**Comment by nonhermitian** [comment]
Great. Tired of getting aer warnings when Aer doesn't even build on the platform.

**Comment by mtreinish** [comment]
Ugh, this is failing the tutorials because ignis is using `from qiskit import Aer` to determine if Aer is available: https://github.com/Qiskit/qiskit-ignis/blob/master/qiskit/ignis/verification/topological_codes/fitters.py#L29-L34 which I wasn't able to preserve with this approach. To unblock I'll have to update the ignis code and switch CI to use ignis master (and we'll have to co-release ignis and terra)

**Comment by mtreinish** [comment]
Actually now I'm not sure what's going on. That line is wrong (per the upgrade notes) but it's not clear why it's using BasicAer (which is what I'm assuming is the cause of the timeouts) there. Aer is installed in the tutorial job and `from qiskit import Aer` will always work with this PR so it should be using Aer. The issue we had in the tests with that for a skip condition was that CI jobs without aer installed would fail because the skip condition was based on `from qiskit import Aer` raising an importerror which it doesn't anymore

File: releasenotes/notes/lazy-load-provider-factory-bdfb3925a2514ef7.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+---
+upgrade:
+  - |
+    The ``qiskit.Aer`` and ``qiskit.IBMQ`` top level attributes are now lazy
+    loaded. This means that the objects will always now exist and warnings will
+    no longer be raised on import if ``qiskit-aer`` or ``qiskit-ibmq-provider``
+    are not installed (or can't be found by Python). If you were checking for
+    the presence of ``qiskit-aer`` or ``qiskit-ibmq-provider`` using these
+    module attributes and explicitly comparing to ``None`` or looking for the
+    absence of the attribute this no longer will work because they are always
```
Comment (kdk) [inline releasenotes/notes/lazy-load-provider-factory-bdfb3925a2514ef7.yaml:10]: Might be worth explicitly pointing out that the `ImportError` if IBMQ or Aer aren't installed will now be deferred from the `import` call to the first attempt to access.

**Review by kdk — APPROVED** [review]
Looks good, one comment on the release note.

File: releasenotes/notes/lazy-load-provider-factory-bdfb3925a2514ef7.yaml (hunk 1 above)
Comment (mtreinish) [reply releasenotes/notes/lazy-load-provider-factory-bdfb3925a2514ef7.yaml:10]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5619/commits/4aac88394a7d4879ccb5368dec051267719290ca

**Review by kdk — APPROVED** [review]

**Pull request merged**